src/ir/var.rs

```diff
@@ -130,6 +130,60 @@ fn default_macro_constant_type(value: i64) -> IntKind {
     }
 }
 
+/// Determines whether a cursor that is pointing into a CXCursor_MacroDefinition
+/// is pointing into a function-like macro. If so, calls the func_macro callback
+/// and returns `Err(ParseError::Continue)` to signal to skip further
+/// processing.  If version to UTF-8 fails (it is performed only where it should
+/// be infallible), then `Err(ParseError::Continue)` is returned as well.
+fn handle_function_macro(
+    cursor: &clang::Cursor,
+    func_macro: impl FnOnce(&str, &[&[u8]]),
+) -> Result<(), ParseError> {
+    // cexpr explicitly does not handle function-like macros, except for a
+    // degenerate case in which it behaves in a non-conformant way. We prevent
+    // cexpr from ever seeing function-like macros by checking for a parenthesis
+    // token immediately adjacent to (that is, abutting) the first token in the
+    // macro definition.
+    let tokens: Vec<_> = cursor.tokens().iter().collect();
+    match tokens.get(0..2) {
+        Some([first, second]) if first.is_abutting(&second) => {
+            // To be doubly sure we are in a function macro definition, we could
+            // check the second macro's type and content, but in order for the
+            // first two macros after `#define` to be adjacent with no
+            // intervening space, the second one has to be an open parenthesis
+            // anyway, so the fact that this code is executing at all implies
+            // that we are defining a function macro.
+            let joined: Vec<_> = tokens
+                .into_iter()
+                .map(|t| t.spelling().to_owned())
+                .collect();
+            let boundary = joined
+                .iter()
+                .position(|s| s == b")")
+                .map(|n| n + 1) // skip past ')'
+                .unwrap_or(joined.len());
+
+            let (left, right) = joined.split_at(boundary);
+            let left = String::from_utf8(left.concat())
```

Comment:
I think the type of the "name" should match the existing functions :
```rust
fn str_macro(&self, _name: &str, _value: &[u8]) {}
fn int_macro(&self, _name: &str, _value: i64) -> Option<IntKind> { None }
```
and for better or for worse, `str_macro` already does this :

https://github.com/rust-lang/rust-bindgen/blob/16189e443c86f06351ef4b8e1c729b67df9d8bc1/src/ir/var.rs#L178

The reason that I chose `&[&[u8]]` rather than `&[&str]` for the expansion is that the expansion could contain (non-UTF8) binary data in strings, I think :
```c
#define MACRO(x) x "\255\255"
char s[] = MACRO(x);
```
I will add a test case that specifically demonstrates this.

Reply:
Moreover, I think we have to respect `will_parse_macro(name: &str)`, right ?

Reply:
I tried to add a test case for a macro like this :
```c
#define TESTMACRO_STRING_FUNC(x) (x "\377\xff")
```
but `cexpr` does not seem to actually interpret the escapes inside the strings, so it does not actually demonstrate what I wanted it to demonstrate.

Reply:
I have not really addressed the "cost" part of your question. I agree that as implemented, the `func_macro` costs CPU cycles even for users that do not need it.

I think that is true for `str_macro` too. Also, `will_parse_macro()` gives one way to avoid some computation.

I will see if I can create a benchmark.

Reply:
I did not think of a good way to create an actual benchmark, but I believe I have reduced the extra tokenization. My changes are a bit invasive, though.

Reply:
Should I try to find a way to memoize the `from_utf8` call ?

Reply:
After reflection, I do not think anyone will pay any extra costs from `from_utf8` -- the call to `String::from_utf8` inside `handle_function_macro` will only occur after it is determined that a function-macro definition is occurring -- so nothing is thrown away.

I regret not thinking of this before putting so many words in this thread.